### PR TITLE
Feature/competitions design

### LINKED ISSUE #1494: Creating competition with (+) action button can not be cancelled
**Describe the bug**
If you already have a competition and want to create a second one, you have to press the action button with the plus sign (+) to open a dialog to create another competition. This dialog can only be close by pressing "Create Competition" and has not cancel option.
Therefore it's not possible to get back to the list of your own competitions, even if switching the navigation with the tabs and other navigation items. The dialog for creating a new competition stays active in the "mine" tabs until you restart the app.

**Is it on a Production or Development Build?**
production

**To Reproduce**
1. Go to 'Competion'
2. Click on 'Mine'
3. Click the action button (+) to create a new competition (you already need to have one)
4. Try to cancel the creation process.

**Expected behavior**
Have a cancel option (either back navigation in the menu or cancel button)

**What actually happens**
No option to cancel the process is offered.

**Screenshots**
Simulator Screen Shot - iPhone 11 - 2019-09-30 at 09 30 55

**Smartphone (please complete the following information):**
- Device: iPhone X
 - OS: iOS 13
 - Plant-for-the-Planet App Version 1.1.0

### PR DESCRIPTION
**Pull request by harshvitra**
---- updated Nov 13----
- [x] scroll layout up when showing the keyboard (e.g. with `KeyboardAwareScrollView`) to allow to see what is entered into fields

---- updated Nov 9----

- [x] Fix buttons on iOS
Screen Shot 2019-11-09 at 3 24 27 PM

- [x] Remove redundant titles from Competition/ Edit/ Add pages.

Screen Shot 2019-11-09 at 3 27 45 PM
Screen Shot 2019-11-09 at 3 28 07 PM

also in edit

- [ ] remove those from translation files.

---- updated Nov 8----

- [x] Competition End date should not be older than Now.  [fix form validation]
Screen Shot 2019-11-08 at 1 14 32 PM

All competition end at midnight.

- [x] Add asterisk if a field is required.
eg: `Tree Target *`
Screen Shot 2019-11-08 at 1 07 21 PM

- [x] In Edit Competition, When a competition without description is `Saved`,  following error shows up.
Screen Shot 2019-11-08 at 1 01 13 PM

- [x] In `Edit Competition` Competition can't be saved if it does not have an image, or description.

- [x] Deleting a competition appears to fail, but actually deletes. (Need to better handle API response). 

Screen Shot 2019-11-08 at 1 05 20 PM

- [x] In Edit Competition , `Who can join is not populated`
Screen Shot 2019-11-08 at 12 25 05 PM

- [x] When the competition is created, the buttons should be populated after checking whether the owner is enrolled or not.

Screen Shot 2019-11-08 at 12 19 24 PM

When the Join button is clicked, the app logs out the current user, and redirects to `World` Page.

- [x] In participant list
Screen Shot 2019-10-23 at 3 32 47 PM

---- updated Nov 6----

- [x] Background is missing on the Delete button.

- [x] Also Selection of `Who can Join` is not being saved.

- [x] You can’t create/edit a competition because of the null error in iOS

- [x] We should show a warning message after delete competition.
Screen Shot 2019-11-07 at 1 47 12 PM

- [x] Fix issue with Image Picker crashing one iOS

---- updated Nov 6----

- [x] In Edit Competition. offer an option to  `Delete` or `Save` instead of the button (Edit Competition) with regular design

After fixing this everything else looks good for v1.2 release.

---- updated Nov 1----

todo

- [x] fix form schema on iOS, dropdown and calendar are broken.

- [x] List (All) should only show competitions which do not have end date before today.

- [x] List (Archived) should only show competitions that have end date before today.

- [x] Option to leave/join for Competition Owner.

- [x] Implement New Form Schema and Layout for Edit Competition Page

New designs for Competitions 

To Do - 

- [x] Implement Form Schema with Material Design

Fix #1494

### CODE REVIEW

**Comment by sagararyal** [comment]
@harshvitra 

this is a conditional field.

If the competition has a project selected.

Screen Shot 2019-09-25 at 11 49 51 AM

**Comment by sagararyal** [comment]
- [ ]  if end date is in the past. Hide the join button and do not show them in Featured and All.
Screen Shot 2019-09-25 at 11 57 13 AM

- [ ] creating competition does not work (could be a form schema thing.

Screen Shot 2019-09-25 at 11 59 27 AM

- [ ] Add a new tab called Archive which shows old competitions.
Screen Shot 2019-09-25 at 12 00 16 PM

- [ ] Different header texts
for All : List of all competitions competitions. Join one, and start planting.
for Archive : List of competitions from the past. 
for Mine: List of all competitions you've joined or created.  If empty, : It seems you don't have any competitions yet. Please join or create one, and they will appear here.
 
Screen Shot 2019-09-25 at 12 01 38 PM

**Comment by sagararyal** [comment]
Further todo:

- [ ] In participant list
Screen Shot 2019-10-23 at 3 32 47 PM

- [ ] fix the form (layout and schema)

- [ ] Fix Edit Competition Screen

Few more include (adding Join/Leave button for competition owner)

**Comment by norbertschuler** [comment]
I reverted the commit a065cbd on behalf of @harshvitra .

**Comment by norbertschuler** [comment]
This pull request require some fixes from eslint. @harshvitra you can look at the other code on how we fixed this stuff e.g. the `react-native/no-color-literals`.

**Comment by harshvitra** [comment]
Apart from Save/Delete rest things are done @sagararyal , date picker and dropdown changed to work for iOS also please check if image picking is working in iOS, I'm not 100% sure about it working in iOS

**Comment by norbertschuler** [comment]
@harshvitra my issue was just about deleting my own competition where I was the only participant.

File: app/components/Competition/Tabs/formComponents.native.js [hunk 1]
```diff
@@ -0,0 +1,275 @@
+import React, { useState } from 'react';
+import { Text, View, Image, TouchableOpacity } from 'react-native';
+import { cameraSolid, imageGallery, forward } from '../../../assets';
+import styles from '../../../styles/competition/competition-form.native';
+import { formatDateToMySQL } from './../../../helpers/utils';
+import { formatDate } from './../../../utils/utils';
+import DateTimePicker from 'react-native-modal-datetime-picker';
```
Comment (norbertschuler) [inline app/components/Competition/Tabs/formComponents.native.js]: I would like to know why the formerly used DatePicker isn’t suitably for your purpose and this new module needs to get used. And is it localized?

File: app/components/Competition/Tabs/formComponents.native.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+import i18n from '../../../locales/i18n';
+import { Formik } from 'formik';
+import { TextField } from 'react-native-material-textfield';
+import competitionFormSchema from './../../../server/formSchemas/competition';
+import { generateFormikSchemaFromFormSchema } from '../../../helpers/utils';
+import ImagePicker from 'react-native-image-crop-picker';
```
Comment (norbertschuler) [inline app/components/Competition/Tabs/formComponents.native.js]: I would also like to know why the formerly use ImagePicker module has to get replaced with this one and if it also get localized? Also I hope the new module can handle permission management for native apps as the old one.

Comment (enroot-mumbai) [reply app/components/Competition/Tabs/formComponents.native.js]: I've checked the permissions part, this works for android, unable to test for iOS for now. 
Reason for integrating this  ImagePicker module is that it has support for picking multiple images which the earlier library isn't very supportive of.
Which we need in Reviews too, had integrated there first and then added it here.

File: app/components/Competition/Tabs/formComponents.native.js (hunk 1 above)
Comment (enroot-mumbai) [reply app/components/Competition/Tabs/formComponents.native.js]: I'll check the localization... 
The module react-datepicker is not for React Native apps, it is just for the web

File: app/components/Competition/Tabs/formComponents.native.js (hunk 2 above)
Comment (harshvitra) [reply app/components/Competition/Tabs/formComponents.native.js]: I've checked the permissions part, this works for android, unable to test for iOS for now.
Reason for integrating this ImagePicker module is that it has support for picking multiple images which the earlier library isn't very supportive of.
Which we need in Reviews too, had integrated there first and then added it here.

File: app/components/Competition/Tabs/formComponents.native.js (hunk 1 above)
Comment (harshvitra) [reply app/components/Competition/Tabs/formComponents.native.js]: react-datepicker is only for web and not for Native apps. I'll check about the localization

Comment (norbertschuler) [reply app/components/Competition/Tabs/formComponents.native.js]: Right, we use https://github.com/Plant-for-the-Planet-org/treecounter-app/blob/98e2a00540281ccf178bf36ea97d608f56e59577/app/components/Templates/DatePickerTemplate.android.js#L8 for Android and https://github.com/Plant-for-the-Planet-org/treecounter-app/blob/98e2a00540281ccf178bf36ea97d608f56e59577/app/components/Templates/DatePickerTemplate.ios.js#L8 for iOS.

File: app/components/Competition/Tabs/formComponents.native.js (hunk 2 above)
Comment (norbertschuler) [reply app/components/Competition/Tabs/formComponents.native.js]: Please also check for localization. I added lots of resources for the former solution which you commented now, but haven’t removed yet from the translation resource files.

File: app/components/Competition/Tabs/formComponents.native.js (hunk 1 above)
Comment (harshvitra) [reply app/components/Competition/Tabs/formComponents.native.js]: Screen Shot 2019-11-06 at 11 58 51 PM

https://facebook.github.io/react-native/docs/datepickerios

Please take this into consideration

Comment (harshvitra) [reply app/components/Competition/Tabs/formComponents.native.js]: The mentioned [DateTimePicker](https://github.com/react-native-community/react-native-datetimepicker)  has issues, I had installed it before and it was crashing on iOS so removed it

And hence this - https://github.com/mmazzarolo/react-native-modal-datetime-picker

**Comment by harshvitra** [comment]
These two things are not ready yet, and we will be working on these in the future. 

- [ ] If the competition has Project selected show card

- [ ] If it has recommended project take user there

**Comment by harshvitra** [comment]
> These two things are not ready yet, and we will be working on these in the future.
> 
> * [ ]  If the competition has Project selected show card
> * [ ]  If it has recommended project take user there

@sagararyal can you test the rest of the PR.

File: app/components/Competition/Tabs/formComponents.native.js (hunk 2 above)
Comment (harshvitra) [reply app/components/Competition/Tabs/formComponents.native.js]: I have removed `react-native-image-crop-picker` for it to work perfectly we need latest React Native versions, so I have unlinked and removed it. Using the `react-native-image-picker `library now.

**Comment by harshvitra** [comment]
> @harshvitra my issue was just about deleting my own competition where I was the only participant.

@norbertschuler this also fixes #1493 now you can delete own competitions.

File: app/components/Competition/Tabs/formComponents.native.js (hunk 1 above)
Comment (norbertschuler) [reply app/components/Competition/Tabs/formComponents.native.js]: Ok, thank you for researching this. Now is the new package localized, i.d. working in German and English (e.g. for months name or date format). I have not yet run the code.

**Comment by norbertschuler** [comment]
> @norbertschuler this also fixes #1493 now you can delete own competitions.

I just created a competition on `staging` and deleted it again, but it was still shown in the app in the mime tab - the API requests look o.k.:

Screenshot_1573304330

Bildschirmfoto 2019-11-09 um 14 01 03

**Comment by norbertschuler** [comment]
I made another test creating a competition and I think the button "Join competition" is wrong as you can see I am already a participant after I created the competition - I am also returned from the API call getting the newly created competition:

Screenshot_1573306393

Bildschirmfoto 2019-11-09 um 14 36 46

**Comment by norbertschuler** [comment]
@harshvitra please merge develop into this branch and fix the merge conflicts.

**Comment by Plant-for-the-Planet** [comment]
@norbertschuler 
this issue will be fixed from backend side, (we are auto enrolling competition owners at the moment, and will change as soon as we release these builds.
Screen Shot 2019-11-10 at 6 05 36 PM


This branch is good to go.

File: app/components/Competition/Tabs/formComponents.native.js (hunk 1 above)
Comment (norbertschuler) [reply app/components/Competition/Tabs/formComponents.native.js]: An Android the date picker is exactly the same as DatePickerAndroid and also localized. For iOS I have not yet tested it.

Comment (norbertschuler) [reply app/components/Competition/Tabs/formComponents.native.js]: On iOS localization is missing: https://www.npmjs.com/package/react-native-modal-datetime-picker

Comment (norbertschuler) [reply app/components/Competition/Tabs/formComponents.native.js]: added the missing localization

**Comment by norbertschuler** [comment]
One more thing: Every time I open the page to add a new competition, I recognized that 6 API calls to tree counters are made which takes some time and within this time the user cannot use the date picker or the image picker or the dropdown:

Simulator Screen Shot - iPhone 11 - 2019-11-11 at 09 47 26

These are the tree counters loaded for me (they match the tree counters I am following) - why are they loaded and why does it block some parts of the UI?

Bildschirmfoto 2019-11-11 um 09 49 45

**Comment by norbertschuler** [comment]
Just a comment: I cannot test the camera with the image picker - in the console you get a comment `ImagePicker Error:  – "Camera not available on simulator"` - so we will have to test this on a real device.

**Comment by norbertschuler** [comment]
The "plant" button got cut in my iPhone 11 simulator:

Simulator Screen Shot - iPhone 11 - 2019-11-11 at 09 57 19

**Comment by norbertschuler** [comment]
If you edit a competition on a real device and want to search someone to invite the keyboard hides the search field and results to choose:

IMG_6597

**Comment by harshvitra** [comment]
> * [ ]  explain loading the tree counter of all followers when creating or editing a competition (does not scale with 100+ followed counters and blocks UI)

This particular API can be hit by - treecounterLookupAction in the actions folder, but I dont see any call made to this function anywhere near Competitions.

**Comment by norbertschuler** [comment]
> > * [ ]   explain loading the tree counter of all followers when creating or editing a competition (does not scale with 100+ followed counters and blocks UI)
> 
> This particular API can be hit by - treecounterLookupAction in the actions folder, but I dont see any call made to this function anywhere near Competitions.

Maybe the error was already there before. I have seen that all the tree counters of follower are loaded whenever you open a tree counter of someone. I filed this as a new bug #1631.

File: app/components/Competition/Tabs/formComponents.native.js [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
+      <DateTimePicker
+        isVisible={showDatePicker}
+        onConfirm={date => {
+          (date = date || props.endDate),
+            setShowDatePicker(false),
+            props.setFieldValue('endDate', date);
+        }}
+        onCancel={() => setShowDatePicker(false)}
+        minimumDate={new Date()}
```
Comment (norbertschuler) [inline app/components/Competition/Tabs/formComponents.native.js]: - [ ] The API expects the minimum date to be at least today + 1:

Simulator Screen Shot - iPhone 11 - 2019-11-14 at 09 25 47

**Comment by norbertschuler** [comment]
Two more issues:

- [ ] deleting a competition deletes it in the backend, but still does not remove it from the frontend - if you had two competition then after deletion the remaining one is shown twice:

Simulator Screen Shot - iPhone 11 - 2019-11-14 at 10 01 10

We either should be able to adapt the local storage correctly when changing (adding, editing or deleting) data or completely refresh it from the backend if we cannot handle the local storage changes.

- [ ] After you created a new competition we show the new competition in the same dialog. If you click the back arrow you will be shown again the input dialog where you just added you new competition. We rather should redirect the user to the list of his own competitions again to avoid confusion with the navigation:

Simulator Screen Shot - iPhone 11 - 2019-11-14 at 10 04 58

**Comment by harshvitra** [comment]
I've solved 4 out of the 5 issues, regarding the delete competition issue. 

What we need to do is - Unlink and Delete entity. 
This is the server response - 
```
{
    "delete": {
        "competition": [
            84
        ]
    },
    "unlink": {
        "treecounter.10.competitions": [
            84
        ]
    }
}
```

Delete happens successfully, but unlink doesn't work. Because of which we get an error as the competition in the object has value of undefined.

**Comment by norbertschuler** [comment]
> Delete happens successfully, but unlink doesn't work. Because of which we get an error as the competition in the object has value of undefined.

Can not help you here as I don't know how local data handling in react-native works. Please ask in our developer Slack channel for a solution.

**Comment by norbertschuler** [comment]
I like you geeky solution to add a day, but I still managed to enter a date before tomorrow - the request send is as following:
```
{
  "name": "6",
  "goal": "6",
  "description": "6",
  "access": "immediate",
  "endDate": "2019-11-14T14:39:37.690Z",
  "imageFile": ""
}

```
I tried myself to replace you code with `minimumDate={addDays(new Date(), 1)}` by importing a helper function with `import addDays from 'date-fns/addDays'`, but even with that I get a request like this:
```
{
  "name": "6",
  "goal": "6",
  "description": "6",
  "access": "immediate",
  "endDate": "2019-11-14T23:00:00.000Z",
  "imageFile": ""
}
```
Perhaps you can investigate what else is done with the date before sending it? Some timezone issue maybe?

**Comment by harshvitra** [comment]
> I like you geeky solution to add a day, but I still managed to enter a date before tomorrow - the request send is as following:
> 
> ```
> {
>   "name": "6",
>   "goal": "6",
>   "description": "6",
>   "access": "immediate",
>   "endDate": "2019-11-14T14:39:37.690Z",
>   "imageFile": ""
> }
> ```
> 
> I tried myself to replace you code with `minimumDate={addDays(new Date(), 1)}` by importing a helper function with `import addDays from 'date-fns/addDays'`, but even with that I get a request like this:
> 
> ```
> {
>   "name": "6",
>   "goal": "6",
>   "description": "6",
>   "access": "immediate",
>   "endDate": "2019-11-14T23:00:00.000Z",
>   "imageFile": ""
> }
> ```
> 
> Perhaps you can investigate what else is done with the date before sending it? Some timezone issue maybe?

I have question, how did you enter the current data, the datepicker did not block the previous date ? The point of note using the date-fns was to avoid the timezone issue by adding 24hrs anyhow and it also take cares of the last day of month and not gives us 32 as date.

File: app/components/Competition/CompetitionParticipant.native.js [hunk 4]
```diff
@@ -1,33 +1,34 @@
...
     this.plantButton = this.plantButton.bind(this);
   }
 
+  // This function is for people to support someone who is a participant in the competition
   supportButton() {
     let supportObject = {
       id: this.props.competitor.treecounterId,
       displayName: this.props.competitor.treecounterDisplayName
     };
+    console.log('support button pressed');
```
Comment (sagararyal) [inline app/components/Competition/CompetitionParticipant.native.js]: do we need this?

Comment (norbertschuler) [reply app/components/Competition/CompetitionParticipant.native.js]: At least all console.log get stripped out for production automatically, but you're right, we should delete them if we don't need them any more.

**Comment by norbertschuler** [comment]
> I have question, how did you enter the current data, the datepicker did not block the previous date ? The point of note using the date-fns was to avoid the timezone issue by adding 24hrs anyhow and it also take cares of the last day of month and not gives us 32 as date.
> 
> I cant pick any date before tomorrow

I just committed the form or used the dropdown to select the current date. Nothing more special. Maybe you can have a look at the other date picker implementation to compare how they got used. I think we always formatted the returning value before passing it to the backend: https://github.com/Plant-for-the-Planet-org/treecounter-app/blob/4132d4d0a92704fc498c1e656f6cd394719cfa3c/app/components/Templates/DatePickerTemplate.js#L32 or https://github.com/Plant-for-the-Planet-org/treecounter-app/blob/4132d4d0a92704fc498c1e656f6cd394719cfa3c/app/components/Templates/DatePickerTemplate.ios.js#L29 or https://github.com/Plant-for-the-Planet-org/treecounter-app/blob/4132d4d0a92704fc498c1e656f6cd394719cfa3c/app/components/Templates/DatePickerTemplate.android.js#L56

**Comment by norbertschuler** [comment]
I can only test on Android today and I got a crash trying to add a new competition with todays date:
```
TypeError: date.getDate is not a function. (In 'date.getDate()', 'date.getDate' is undefined)
This error is located at:
    in CompetitionDatePicker (at formComponents.native.js:93)
    in RCTView (at View.js:45)
    in View (at formComponents.native.js:73)

```

**Comment by norbertschuler** [comment]
> Delete happens successfully, but unlink doesn't work. Because of which we get an error as the competition in the object has value of undefined.

Deletion works for me now :-) But for competitions which already ended, the user cannot edit and therefore also do not delete them anymore as the UI does not allow it. I do not know what the concept says about this use case.

**Pull request merged**